### PR TITLE
[fx] rewrite `FloorDiv` to match Python better

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91376
* __->__ #90906

### CODE REVIEW

**Comment by nkaretnikov** [comment]
There are some tests in `test/test_dynamic_shapes.py` for `floordiv` and `float`, so I didn't add more.

Example of incorrect behavior: `FloorDiv` doesn't support this (it just returns `base` as is):
```
In [54]: 2.5 // 1
Out[54]: 2.0
```

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:260]: perhaps not creating these strings is better and just throw a generic error msg?

**Comment by nkaretnikov** [comment]
@ezyang the functorch failure is real, however, I'm not quite sure what to do there:

The issue is in `slice_forward` decomp: `sizes[dim] = (len + step - 1) // step`, which could be: `s0 - 2*s0//3 1`.

Even if I add a check for `and not isinstance(x, SymInt) and not isinstance(x, sympy.Integer)` and wrap both arguments into `sym_int` at call site, it doesn't help it.

Is there currently a way to check that the value of a symbolic expression is an integer? I'm assuming we could potentially track it if we do not do FP ops, right? But we just don't?

Basically, no way to have this checked in `FloorDiv` now, correct? Or... does it mean that we actually need to implement `FloorDiv` properly?

**Comment by ezyang** [comment]
This failure is surprising to me because it seems like all the arguments to floor div should be ints...

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:256]: This isinstance check looks very questionable. I think the real problem here is that we didn't plug in FloorDiv into sympy's assumptions system; if both inputs to floor div are integers, the output should be an integer too.

cc @Chillee

**Comment by ezyang** [comment]
I don't see any call to slice in the test case you quoted?

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:280]: OK. What I expect to see is that if we have bugs in assumption propagation, we will incorrectly fire TypeError. I think the correct fix is to get assumption propagation fixed. But since testing for complex inputs is very niche, I think it would also be OK to delete this logic if necessary.

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:297]: skip the sympy.floor if base.is_integer? I guess maybe sympy does that automatically

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:311]: I'm not sure we want to actually do this if the base and divisor are integer; I think I would rather keep the floor div unevaluated in this case

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:297]: Yes, it depends on the type. For Integer, it just returns: https://github.com/sympy/sympy/blob/master/sympy/core/numbers.py#L2121

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: Done, but note that I also had to update `evaluate_expr` because of it. We can no longer short-circuit there. Otherwise, an uneval'd `FloorDiv` will propagate and will fail somewhere else. E.g., in `SymNode.guard_float` when it tries to do `float(FloorDiv(...))`.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py]: I did somehow miss `is_complex` before, so this is removed.

**Comment by ezyang** [comment]
cc @asmeurer

**Comment by asmeurer** [comment]
If it's of any help, I recently wrote a new guide in the SymPy docs on how to define custom functions https://docs.sympy.org/latest/guides/custom-functions.html.

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1023]: I don't quite understand. Are you saying something like `FloorDiv(2, 3)` won't automatically simplify? Does that seem like a problem with the built-in simplifier, it should simplify past this no?

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:277]: I have trouble telling whether or not and/or have higher precedence, so best to parenthesize more here (in contrast, the outer parenthesis is completely unnecessary)

Comment (asmeurer) [reply torch/fx/experimental/symbolic_shapes.py:277]: See the third bullet point here https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval. I recommend not failing here if assumptions are None.

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:269]: I guess my problem with this comment is that it reads like "well, I tried a bunch of things and they didn't work so here's the thing we're doing now" but the comment never lays out what it is exactly that you're trying to do, so an uniformed reader will look at the comment and be like, "ok, so sympy represents complex with add exprs, why do I care".

But Sympy in fact gives *explicit* guidance about what to do when defining functions restricted to certain domains, e.g., https://docs.sympy.org/latest/guides/assumptions.html#implementing-assumptions-handlers so really we should quote the docs and then follow what they say there. (BTW, it doesn't look like your code does what the docs say to do lol)

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1023]: To give a concrete example, with the above two lines reverted:

```
% pytest test/test_dynamic_shapes.py -k floordiv -v
AssertionError: Object comparison failed: 1.0 != 2.9//2.5
```

Given all the suggestions here and in the docs, we probably want to leave some things uneval'd. I'll see if I can keep this while preserving correct behavior.

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1023]: Ah, if we do not fully simplify in `evaluate_expr` (as done in this PR), we may end up in a situation where we try to do `guard_int/float` on a `FloorDiv`. That will fail because those are defined like this: `float(self.shape_env.evaluate_expr(self.expr))`. Which will simplify to `float(FloorDiv(a, b))` and fail. We even have a test for this in `TestSymNumberMagicMethods`.

File: test/test_dynamic_shapes.py
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:627]: Missed this, will fix.

File: test/test_dynamic_shapes.py
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:724]: This is a new test for the assumptions-related changes.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:248]: Rewrote the note with some context + new sympy links.

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:248]: This should probably also mention how we check for types. That is, we allow for None.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:277]: Changes here are just due to me introducing properties. I think this is more readable.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:292]: This now allows for None.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: Updated the comment and added things specific to the assumptions system.

I still need to rebase and clean up minor issues, but feedback welcome.

**Comment by nkaretnikov** [comment]
@ezyang CI is green, PTAL.

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1038]: Do the len(expr.free_symbols) test here then? But I still think there's something fishy going on here. How come everything else in sympy was actually fully evaluated at this point without simplify...

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1038]: - `simplify` won't work because we don't define some cases for our `FloorDiv`, it won't know what to do:
```py
[ins] In [251]: class Moo(sympy.Function):
           ...:     @classmethod
           ...:     def eval(self, x, y):
           ...:         if x < 2:
           ...:             return x + y
           ...:

[ins] In [252]: Moo(2,3)
Out[252]: Moo(2, 3)

[ins] In [253]: Moo(1,3)
Out[253]: 4

[ins] In [254]: sympy.simplify(Moo(2,3))
Out[254]: Moo(2, 3)
```

- that's by design: we're subclassing `Function` which is for symbolic exprs (AKA exprs you want to keep uneval'd in some cases)

- some things in sympy subclass `Function` too, but many subclass `Expr` instead, like `Add`. Those that are symbolic will fail to simplify too:

```py
[ins] In [263]: sympy.simplify(sympy.log(sympy.sympify(8)/3, 2))
Out[263]: -log(3)/log(2) + 3
```

- depending on how you define eval, the advantage of symbolic is that you can simplify without compute or just with Symbols, but you still need to eval somewhere. we eval in `_maybe_evaluate_static` by pattern-matching on `FloorDiv`

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1038]: bumping this comment

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1038]: Aha! So `simplify` also automatically calls `doit`, which you can define on a class for full eval. There we can eval and return after `simplify` as before if there are no free vars. I'll try implementing that.

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1052]: Added doit to have more control over eval + some tests. PTAL

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1052]: Also: The `log` snippet posted above in this thread has no free vars either. And it's not fully eval'd. So the eval strategy is up to however you want it to work.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1052]: That's because Moo is not total. All of the functions we are defining are total

**Review by ezyang — CHANGES_REQUESTED** [review]
I don't think doit should be necessary. I will take an attempt at modifying the PR

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1052]: Right, but Moo is essentially FloorDiv. Note that we sometimes return None from FloorDiv as well. Which is why we need to fully eval it somewhere else.

Feel free to ping me offline if you want me to explain further.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1052]: FloorDiv should be total when all the inputs are Integer (aka known literals)

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1052]: It does simplify in the Integer case as is:

```
[nav] In [15]: FloorDiv(6, (FloorDiv(12, 6)))
Out[15]: 3
```

The reason _additional_ eval is needed is that `FloorDiv` might return None in some cases (when the `eval` classmethod returns None). So we need to somehow convert `FloorDiv` to an actual value in `evaluate_expr`. Using `doit` is one way of doing that.

I also mentioned it before, but the reason we need to return a concrete value from `evaluate_expr` is that it's used in `guard_float`/`guard_int`. E.g., `guard_int` will attempt to call `int(FloorDiv(...))`, which won't work.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1052]: It should simplify even with mixed arguments

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1052]: I suppose one could hypothetically argue there would be precision could be lost for some ops so they shouldn't evaluate, but I don't feel this problem applies to floordiv

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1052]: > It should simplify even with mixed arguments

But I don't see how this could work. Either:
- we need to make FloorDiv total for all inputs (it was mentioned before that we don't want to do that)
- OR we could move the code I have in doit into FloorDiv and apply it for LHS and RHS sub-exprs (eval if no free vars).

Are you okay with either of these solutions? Or do you see a different way to approach this?

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1052]: It seems clear to me that FloorDiv's eval function should actually evaluate when the inputs are Integers or Floats. It seems crazy to not evaluate with a Float inputs.

Why do you think we don't want to make FloorDiv total for this case? Are you talking about your comment https://github.com/pytorch/pytorch/pull/90906#discussion_r1064048676 ? I reread the Sympy documentation on the matter and it doesn't seem to say anything about the Integer/Floating input case.

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1052]: - I can add this code to `FloorDiv.eval`, then it'll eval Float/Integer inputs. Problem solved. _I'll push this fix shortly._
```py
if isinstance(base, (sympy.Integer, sympy.Float)) and isinstance(divisor, (sympy.Integer, sympy.Float)):
    return sympy.floor(base / divisor)
```

- I was concerned about funcs that _have no free vars_ and _do not simplify fully_. E.g., `log`:
`sympy.log(sympy.sympify(8)/3, 2)`. Turns out, however, it doesn't simplify even with `.doit(deep=True)`, so no point in keeping our `doit` code around, I guess. I might be wrong, but I feel like these should be pretty rare, because most arithmetic funcs are not defined by wrapping Function in sympy. I think we can address this later if there are problems.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:290]: This is a separate case to avoid calling floor when both args are Integers.

File: test/test_dynamic_shapes.py
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:700]: All eval to 3 now.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: Updated. LMK if there's anything else that needs changing. Otherwise, please stamp it so I could merge. (I'll re-read the whole diff once again before merging to make sure we didn't miss any minor issues.)

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:290]: Okay, I'll merge these two PRs after review as is to avoid stressing the CI for no good reason, then will fix this in a follow up.

File: torch/fx/experimental/symbolic_shapes.py
Comment (asmeurer) [reply torch/fx/experimental/symbolic_shapes.py]: At the end of the day, the degree of automatic evaluation that you have in `eval` is up to you, and it depends on what you want out of your function. If evaluating too little is an annoyance that makes you have to call `doit` too much, then that may be a sign that you should move the evaluation to `eval` to happen automatically. 

I've tried to outline some of the pitfalls to avoid here https://docs.sympy.org/latest/guides/custom-functions.html#defining-automatic-evaluation-with-eval, but even if you don't follow those guidelines, SymPy will still support whatever degree of evaluation you choose to implement. That doc tends to caution towards less automatic evaluation because you can always go in one direction, to evaluate something that doesn't evaluate automatically, but doing the reverse---preventing something from evaluating automatically---is harder (it can only be done with `evaluate=False`, which is fragile). This all comes from the experience of the sorts of issues we've run into in SymPy itself. I do recommend evaluating on explicit numerical inputs (when the value is itself a number) because it's rare to not want that. 

FWIW, I personally think `log` does too much automatic evaluation on rational numbers (unfortunately, quite a few functions in SymPy itself don't exactly comply with the advice given in that document). I would much rather it only evaluated when the result doesn't have any logs, and leave the rest to simplification functions like `expand_log` and `logcombine`.

**Comment by malfet** [comment]
So looks like this PR caused a regression. And I guess the classification for both should be `nosignal` - inductor tests are no run on PR that do not add `ciflow/inductor`, so..

**Comment by nkaretnikov** [comment]
I've been told by Horace in https://github.com/pytorch/pytorch/pull/92397#discussion_r1080645846 that flooring base when divisor is 1 might break things in inductor. So I'm going to enable dynamic shapes tests first as suggested in https://github.com/pytorch/pytorch/pull/91952#discussion_r1070724600 before getting back to this, so we have good coverage.

**Comment by nkaretnikov** [comment]
This hasn't received updates in a while, so let me write down a small note. The model fails because of the `_eval_is_real` and `_eval_is_integer` methods. With these, `requested_length == x` in `_refs.expand` returns False and the check fails.

I'm currently trying to understand what exactly happens there. I think it would be beneficial to keep these if we could make them work.

**Comment by ezyang** [comment]
I am confused. Are you complaining that after you tightened up the assumptions, sympy was able to simplify even more than it used to? That seems like a good thing to me! If we needed to extract equalities from the expression we should figure out some other way to do it then

**Comment by asmeurer** [comment]
I'm also a little unclear on exactly what is going on. One thing that pops out to me is that you replaced floor_div(x, y) with div(x, y), but these are not equal in general. When `FloorDiv(s2 , 8) = s2 / 8` the equality is true, but this is only the case when s2 is a multiple of 8 (you can see that the equality is true in this case by replacing s2 with `8*k` and simplifying). If it is indeed true that `s2` is a multiple of 8 then the resulting answer should be correct.

**Comment by nkaretnikov** [comment]
> I am confused. Are you complaining that after you tightened up the assumptions, sympy was able to simplify even more than it used to? That seems like a good thing to me! If we needed to extract equalities from the expression we should figure out some other way to do it then

Auto simplifications == good, yes. But assumptions + truediv in `ShapeEnv.simplify` is unsound.

We add a rule that says "the result of `FloorDiv` can only be an integer if both args are integers", then we supply `a/b` as one of the arguments to `FloorDiv`, which is always a float. As these are expressions with symbols, SymPy infers that `Eq(FloorDiv(base, a/b), integer_expr)` can never be true because they have different assumptions: `Eq(float_expr, integer_expr)`.

There are two options:
1. Remove assumptions and keep things as is.
2. Add assumptions and change/remove truediv in `ShapeEnv.simplify`.

I'm planning to go with (1) for now. We don't have assumptions on master and (2) is more complicated. This PR is already an improvement compared to what we have on master, so I don't see a good reason to delay this further.

We can do (2) separately in a follow up. Challenges with (2): we need to convince SymPy that the expr will be an integer. But we don't want to use `//` (recursive), `sympy.floor(a/b)` (fails to simplify), `sympy.Integer` (doesn't work on Symbols). So we might need to just remove the truediv part, but that's to be investigated.

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:315]: I'll add a special case for when base is an integer if the CI passes with this

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:1075]: No regression here as there are no significant changes to `FloorDiv`. Only needed because of the new `test_floordiv_simplify` test, which exposed this issue

File: torch/fx/experimental/symbolic_shapes.py
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:1209]: Likewise, only because of new tests that exposed this

**Comment by ezyang** [comment]
Well hang on. Python will happily test equality between an int and a float.

```
>>> 2 == 2.0
True
```

If we are turning == into sympy.Eq, we had better preserve this behavior! If sympy says sympy.Eq(float, int) is always false, then we can't use sympy.Eq, we need to roll our own (or maybe, insert the necessary conversions? Converting int to float before this equality test seems like it would work too)

But also, this a/b into FloorDiv case seems really contrived, because nearly ~all of the time I expect the arguments to FloorDiv to be integers.

**Comment by asmeurer** [comment]
> If we are turning == into sympy.Eq, we had better preserve this behavior! If sympy says sympy.Eq(float, int) is always false, then we can't use sympy.Eq, we need to roll our own (or maybe, insert the necessary conversions? Converting int to float before this equality test seems like it would work too)

Currently SymPy compares ints and floats equal but aren't equal in the assumptions. Is this is the source of the problem:

```py
>>> sympy.Float(1.0) == sympy.Integer(1)
True
>>> sympy.Float(1.0).is_integer
False
```

We'd like to clean this up so that they aren't equal but it's hard to do https://github.com/sympy/sympy/pull/20033. 

Anyway, I thought all this stuff was about symbolic shapes and indices, so shouldn't everything be an integer in the end? SymPy won't give you a float unless you use `evalf` (`/` on SymPy integers returns a `Rational`). `a/b` will only be a float if neither of them is a SymPy object. It might just be you need to use `sympy.S(a)/b` so that you end up with a Rational (see https://docs.sympy.org/latest/tutorials/intro-tutorial/gotchas.html#two-final-notes-and).

**Comment by nkaretnikov** [comment]
This PR finally passes, but there are too many changes. I'm going to split it up and resubmit.

**Pull request closed without merging**
